### PR TITLE
Modify array task behavior based on task type version

### PR DESCRIPTION
**Pull request by katrogan**
# TL;DR
Modify array task input reader behavior based on task type version.

With native typing flytekit, dynamic jobs no longer write individual input files for each array task instance. Now we can eliminate the lookup to `prefix/<position>/inputs.pb` and therefore no longer need to pass the parent directory in a modified input reader.

Also, this change fixes k8s array task log links which were never working?

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/609

## Follow-up issue
_NA_

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -18,7 +18,7 @@ require (
 	github.com/hashicorp/golang-lru v0.5.4
 	github.com/kubeflow/pytorch-operator v0.6.0
 	github.com/kubeflow/tf-operator v0.5.3
-	github.com/lyft/flyteidl v0.18.9
+	github.com/lyft/flyteidl v0.18.13-0.20210223003616-672bde1c92d1
```
Comment (katrogan) [inline go.mod:21]: will update before merging

File: go/tasks/plugins/array/k8s/task.go [hunk 2]
```diff
@@ -75,8 +77,26 @@ func (t Task) Launch(ctx context.Context, tCtx core.TaskExecutionContext, kubeCl
...
+	var inputReader io.InputReader
+
+	taskTemplate, err := tCtx.TaskReader().Read(ctx)
+	if err != nil {
+		return LaunchError, errors2.Wrapf(ErrGetTaskTypeVersion, err, "Unable to read task template")
+	} else if taskTemplate == nil {
+		return LaunchError, errors2.Wrapf(ErrGetTaskTypeVersion, err, "Missing task template")
+	}
+
+	if taskTemplate.Version == 1 {
```
Comment (EngHabu) [inline go/tasks/plugins/array/k8s/task.go:89]: >= ?

Comment (katrogan) [reply go/tasks/plugins/array/k8s/task.go:89]: sorry not following

File: go/tasks/plugins/array/k8s/task.go [hunk 3]
```diff
@@ -75,8 +77,26 @@ func (t Task) Launch(ctx context.Context, tCtx core.TaskExecutionContext, kubeCl
...
+		inputReader = tCtx.InputReader()
+	} else {
```
Comment (kumare3) [inline go/tasks/plugins/array/k8s/task.go:91]: I guess we can default this behavior for taskTemplate.Version == 0, as after 1, there maybe 2 etc, which may not change this behavior?

i.e. IMO we should invert the case

Comment (katrogan) [reply go/tasks/plugins/array/k8s/task.go]: updated @kumare3

File: go/tasks/plugins/array/k8s/task.go [hunk 4]
```diff
@@ -75,8 +77,27 @@ func (t Task) Launch(ctx context.Context, tCtx core.TaskExecutionContext, kubeCl
...
+	var inputReader io.InputReader
+
+	taskTemplate, err := tCtx.TaskReader().Read(ctx)
+	if err != nil {
+		return LaunchError, errors2.Wrapf(ErrGetTaskTypeVersion, err, "Unable to read task template")
+	} else if taskTemplate == nil {
+		return LaunchError, errors2.Wrapf(ErrGetTaskTypeVersion, err, "Missing task template")
+	}
+
+	switch taskTemplate.GetTaskTypeVersion() {
```
Comment (kumare3) [inline go/tasks/plugins/array/k8s/task.go:89]: Switch 😊 overkill

**Review by kumare3 — DISMISSED** [review]

File: go/tasks/plugins/array/k8s/task.go (hunk 4 above)
Comment (katrogan) [reply go/tasks/plugins/array/k8s/task.go]: I thought we were concerned about future-proofing! updated

**Comment by katrogan** [comment]
PTAL @kumare3 @EngHabu

File: go/tasks/plugins/array/k8s/task.go [hunk 5]
```diff
@@ -75,8 +77,26 @@ func (t Task) Launch(ctx context.Context, tCtx core.TaskExecutionContext, kubeCl
...
+	var inputReader io.InputReader
+
+	taskTemplate, err := tCtx.TaskReader().Read(ctx)
+	if err != nil {
+		return LaunchError, errors2.Wrapf(ErrGetTaskTypeVersion, err, "Unable to read task template")
+	} else if taskTemplate == nil {
+		return LaunchError, errors2.Wrapf(ErrGetTaskTypeVersion, err, "Missing task template")
+	}
+
+	if taskTemplate.GetTaskTypeVersion() == 0 {
```
Comment (EngHabu) [inline go/tasks/plugins/array/k8s/task.go:89]: Should you replicate this for AWS Batch plugin?

Comment (katrogan) [reply go/tasks/plugins/array/k8s/task.go:89]: yes, thank you! done @EngHabu

File: go/tasks/plugins/array/k8s/task.go (hunk 2 above)
Comment (EngHabu) [reply go/tasks/plugins/array/k8s/task.go]: :-D Sorry meant to say ">=" but that got converted into a quote :-D you addressed that anyway

**Comment by kumare3** [comment]
I will let @EngHabu comment, I think I do not have enough context in this code base. 
Also probably @anandswaminathan and @migueltol22

File: go/tasks/plugins/array/k8s/monitor.go [hunk 6]
```diff
@@ -112,6 +112,7 @@ func LaunchAndCheckSubTasksState(ctx context.Context, tCtx core.TaskExecutionCon
 
 		var monitorResult MonitorResult
 		monitorResult, err = task.Monitor(ctx, tCtx, kubeClient, dataStore, outputPrefix, baseOutputDataSandbox)
+		logLinks = task.LogLinks
```
Comment (EngHabu) [inline go/tasks/plugins/array/k8s/monitor.go:115]: needed?

**Review by EngHabu — DISMISSED** [review]
LGTM. Let's wait to merge until we have tested end to end though?

File: go/tasks/plugins/array/k8s/monitor.go (hunk 6 above)
Comment (katrogan) [reply go/tasks/plugins/array/k8s/monitor.go:115]: yes even though everything i read indicates golang slices are effectively pass by reference this failed on sandbox deployment unless i updated

**Comment by katrogan** [comment]
@EngHabu what do you envision by end to end testing? I've tested this locally on my sandbox with task_type_version = 1 unless do you mean adding a full functional test to the end to end test check?

**Review by EngHabu — APPROVED** [review]
